task-maker-async/src/file_set.rs

```diff
@@ -0,0 +1,265 @@
+        }
+        self.kind = kind;
+        self.status = FileSetStatus::Created;
+        self.creation_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        Ok(true)
+    }
+
+    pub fn start_writing(&mut self) -> bool {
+        if self.status > FileSetStatus::Created {
+            return false;
+        }
+        self.status = FileSetStatus::Writing;
+        self.writing_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        true
+    }
+
+    pub fn mark_finalized(&mut self) -> bool {
+        if self.status > FileSetStatus::Writing {
+            return false;
+        }
+        self.status = FileSetStatus::Finalized;
+        self.finalization_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        for file in self.files.iter_mut() {
+            file.1.readers.drain(..).for_each(|mut waiter| {
+                let _ = waiter.send(FileReadingOutcome::from_data(
+                    &file.1.data,
+                    waiter.offset,
+                    waiter.chunk_size,
+                ));
+            });
+        }
+        true
+    }
+
+    pub fn append_to_file(&mut self, file: &FileSetFile, data: &[u8]) {
+        let file_data = self.files.entry(file.clone()).or_default();
+        file_data.data.extend_from_slice(data);
+        file_data.readers.drain(..).for_each(|mut waiter| {
+            let _ = waiter.send(FileReadingOutcome::from_data(
+                &file_data.data,
+                waiter.offset,
+                waiter.chunk_size,
+            ));
+        });
+    }
+
+    pub fn read_from_file(
+        &mut self,
+        file: &FileSetFile,
+        offset: usize,
+        chunk_size: usize,
+    ) -> impl Future<Output = FileReadingOutcome> {
+        let (outcome, receiver) = if !self.files.contains_key(file)
+            && self.status == FileSetStatus::Finalized
+        {
+            (Some(FileReadingOutcome::Dropped), None)
+        } else {
+            let file_data = self.files.entry(file.clone()).or_default();
+            let outcome = FileReadingOutcome::from_data(&file_data.data, offset, chunk_size);
+            if outcome != FileReadingOutcome::EndOfFile || self.status == FileSetStatus::Finalized {
+                (Some(outcome), None)
+            } else {
+                let (sender, receiver) = channel();
+                file_data.readers.push(FileReadWaiter {
+                    sender: Some(sender),
+                    offset,
+                    chunk_size,
+                });
+                (None, Some(receiver))
+            }
+        };
+        async {
+            if let Some(outcome) = outcome {
+                outcome
+            } else {
+                // The sender for this receiver can never be dropped without having received a
+                // message first.
+                receiver.unwrap().await.unwrap()
```

Comment:
What if the file is dropped?

Reply:
Drop for FileSet will send a FileReadingOutcome::Dropped message.